### DISCUSSION TITLE
Passing a marker on a session so that event handlers can associate data?

### DISCUSSION THREAD

**Discussion by cricalix** [sqlalchemy/sqlalchemy#10547] [Usage Questions]
I have some code that essentially parses a file with an Antlr4-based parser, and emits a series of instructions (verb, noun, args).

A piece of dispatcher code receives all of the instructions as input, starts a SQLAlchemy 1.4 async-enabled Session, and invokes handlers based on the verb and noun. The session is stored in an attribute of a class instance - this class has other attributes that are important for the handler functions, and is passed in as an imaginatively named "ctx" (but is not itself a context manager).

Each handler will do some number of session-based "get this data", "add this data", "update this data" actions - session.query(), session.add() etctera are used via the ctx.session argument that was passed in.

A key design factor is that everything from the text file always happens in one transaction, so any failure rolls back at the DB level.

I'd like to measure how many of each type of underlying SQL statements are emitted, and associate them back to the verb+noun pairing (and their runtime) as a way to keep track of the "cost" of any given instruction. It's easy enough to do a 'before_execute' event handler and use the first word of the statement as a clue about which operation type is happening. However, because SQLAlchemy can delay operations without explicit flush/commit actions, code that looks roughly like:
```
ctx = MyContextThing()
ctx.session = get_sqlalchemy_session()
for x in instructions:
  ctx.inserts = 0
  ret = handle_op(x, ctx)
  print(f"{ctx.inserts=}")

def handle_op(x, ctx) -> bool:
  ctx.session.add(Foo(name=x.name))
  return True
```
misses some of the statements that are processed. A given verb+noun might execute 3 inserts, but the actual insert activity doesn't happen until processing the next verb+noun from the list of instructions.

My though is that it'd be useful to pass a marker to the SQLAlchemy session that would let me attribute a given set of SQL operations to that token via the ctx that's available to all the handlers. Done crudely, I want to do something like
```
for x in instructions:
  ctx.inserts = 0
  ctx.marker = some_unique_id()
  ret = handle_op(x, ctx)

def handle_op(x, ctx) -> bool:
  ctx.session.add(Foo(name=x.name), marker=ctx.marker)
  return True
```

and then access that marker value in the event handler code that's listening for "before_execute".

I've also mused whether session events like before_flust or after_flush are the better place to do this, but there's still no way to say "these operations are associated to this verb+noun".

Other thoughts on how I might achieve this goal are appreciated.

**Comment by zzzeek** [comment]
you can build up structures to track things of your own using [session.info](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.info).   in this case if you want to associate tags with specific objects you might add a dictionary or weak-key referencing dictionary to info:

```py
def handle_op(x, ctx) -> bool:
  f1 = Foo(name=x.name)

  # set up "markers", this can also happen ahead of time, such as in after_transaction_create event
  if 'markers' not in ctx.session.info:
     ctx.session.info['markers'] = {}
  ctx.session.info['markers'][f1] = ctx.marker
  ctx.session.add(f1)
  return True
```

**Reply by cricalix** [reply]
Thank you Mike. Haven't forgotten about this; went to do some testing using SQLite as the backing store (faster for simple tests), found some of the test setup I'm working within had explicit MySQL expectations, and spent two weeks rewriting all of that so I could test with SQLite to prove the concept.
